File: javascript/leetcode/path-sum-iii.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} sum
 * @return {number}
 */
var pathSum = function(root, sum) {
    if (!root) { return 0; }
    return findSum(root, sum) + pathSum(root.left, sum) + pathSum(root.right, sum);
};

var findSum = function(root, sum) {
    if (!root) { return 0; }
    return (root.val === sum ? 1 : 0) + findSum(root.left, sum - root.val) + findSum(root.right, sum - root.val);
}

File: ruby/leetcode/path-sum-iii.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @param {Integer} sum
# @return {Integer}
def path_sum(root, sum)
    return 0 if root == nil
    return find_sum(root, sum) + path_sum(root.left, sum) + path_sum(root.right, sum)
end

def find_sum(root, sum)
    return 0 if root == nil
    this_sum = if root.val == sum then 1 else 0 end
    return this_sum + find_sum(root.left, sum - root.val) + find_sum(root.right, sum - root.val)
end
